Rewrite in TypeScript

rafaelramalho19 opened the pull request:
Closes #86

It's my first time working on Typescript, so this can be a bit bad.


<!-- Issuehunt content -->

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#86: Improve the TypeScript types](https://issuehunt.io/repos/61670221/issues/86)
---

IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>
<!-- /Issuehunt content-->

satazor commented on .eslintrc:
```diff
@@ -0,0 +1,4 @@
+{
+    "extends": ["xo", "xo-typescript"],
+    "parser": "@typescript-eslint/parser"
+}
```
Missing new line

satazor commented on .gitignore:
```diff
@@ -1,2 +1,4 @@
 node_modules
 yarn.lock
+.nyc_output
+dist
```
Same

satazor commented on tsconfig.json:
```diff
@@ -0,0 +1,13 @@
+	"include": [
+		"source/*"
+	]
+}
```
Same

papb commented on source/index.ts:
```diff
@@ -1,28 +1,26 @@
-const onetime = require('onetime');
-
-const plainObject = () => Object.create(null);
+import fs = require('fs');
+import path = require('path');
+import crypto = require('crypto');
+import assert = require('assert');
+import EventEmitter = require('events');
+import dotProp = require('dot-prop');
+import makeDir = require('make-dir');
+import pkgUp = require('pkg-up');
+import envPaths = require('env-paths');
+import writeFileAtomic = require('write-file-atomic');
+import Ajv = require('ajv');
+import debounceFn = require('debounce-fn');
+import semver = require('semver');
+import onetime = require('onetime');
```
Is this syntax correct? Shouldn't it be `import fs from 'fs'` and the like?

rafaelramalho19 replied:
Although it does work, it's not the correct way.

Thanks for noticing this 😄

sindresorhus replied:
No, this syntax is actually correct.

sindresorhus replied:
2

rafaelramalho19 replied:
Updated 👍

papb replied:
Oh, I finally [figured it out](https://github.com/sindresorhus/mem/issues/31#issuecomment-475858825), amazing...

papb replied:
@sindresorhus Nowadays do you still recommend the `import foo = require('foo')` syntax for importing CJS-only modules?

Sorry for the clutter, but this was the "less worse" place to ask that I could think of. Thank you.

sindresorhus commented on source/index.ts:
```diff
@@ -39,21 +37,67 @@ const checkValueType = (key, value) => {
+
+	encryptionKey?: string;
+
+	events?: EventEmitter;
```
Why is this optional? Same with some of the others here.

rafaelramalho19 replied:
The only one was actually `_validator` 😅 

Nice catch

sindresorhus commented:
- The doc comments in index.d.ts should be moved to the new index.ts.
- Try to make sure that the output `dist/index.d.ts` matches the old `index.d.ts` as close as possible.
- Don't use `any` or `object`. Use a specific type whenever possible, and if it's supposed to be non-specific, use `unknown`.
- https://github.com/sindresorhus/typescript-definition-style-guide has a lot of useful info about best practices.

sindresorhus commented on source/index.ts:
```diff
@@ -367,65 +416,74 @@ class Conf {
+			// Legacy decryption without initialization vector
+			// tslint:disable-next-line
+			// eslint-disable-next-line node/no-deprecated-api
+			const decipher = crypto.createDecipher(encryptionAlgorithm, this.encryptionKey);
```
Doesn't matter. It's just a legacy fallback.

rafaelramalho19 commented on source/index.ts:
```diff
@@ -367,65 +416,74 @@ class Conf {
-	get size() {
-		return Object.keys(this.store).length;
-	}
+	encryptData(data: any): Buffer | undefined {
```
It actually needs to be

`encryptData(data: string | Buffer): string | Buffer`

Updated it!

rafaelramalho19 commented on package.json:
```diff
@@ -51,14 +81,24 @@
+		"@types/semver": "^6.2.0",
+		"@types/write-file-atomic": "^2.1.2",
+		"@typescript-eslint/eslint-plugin": "^1.11.0",
+		"@typescript-eslint/parser": "^1.11.0",
```
Done :)

rafaelramalho19 commented on test/index.ts:
```diff
@@ -1,36 +1,38 @@
 const fixture = '🦄';
 
-test.beforeEach(t => {
+test.beforeEach((t: any) => {
```
I've tried other solutions. But it seems to all throw errors when using `t.context.config<...>`

Any ideas?

sindresorhus replied:
https://github.com/avajs/ava/blob/master/docs/recipes/typescript.md#typing-tcontext

rafaelramalho19 replied:
Thanks 👍

rafaelramalho19 commented on source/index.ts:
```diff
@@ -39,21 +37,67 @@ const checkValueType = (key, value) => {
+};
+type ConfMigrations = {
+	[key: string]: (store: Conf) => void;
+};
```
Done :)

sindresorhus commented on source/index.ts:
```diff
@@ -224,7 +303,7 @@ class Conf {
 	}
 
-	_containsReservedKey(key) {
+	_containsReservedKey(key: string | object): boolean {
```
```suggestion
	_containsReservedKey(key: string | {[key: string]: unknown}): boolean {
```

rafaelramalho19 replied:
Done it manually, had to update other places :)

rafaelramalho19 commented:
> * The doc comments in index.d.ts should be moved to the new index.ts.
> * Try to make sure that the output `dist/index.d.ts` matches the old `index.d.ts` as close as possible.
> * Don't use `any` or `object`. Use a specific type whenever possible, and if it's supposed to be non-specific, use `unknown`.
> * https://github.com/sindresorhus/typescript-definition-style-guide has a lot of useful info about best practices.
> * Revert: [ed72882#diff-21e9ddd93162651bd36f6e5bbfca8460L1-R14](https://github.com/sindresorhus/conf/commit/ed72882797aac361f35d0c3f567d95d246740b26#diff-21e9ddd93162651bd36f6e5bbfca8460L1-R14)

Regarding:
1️⃣ I also moved the exports to the bottom of the file so it doesn't get in the way of reading the main part of the source code. Is that okay with you?
2️⃣ Done.
3️⃣ I've removed all I could. The ones that remained seem a bit necessary 🤔 
4️⃣ Thanks for the read 👍 
5️⃣ Done, I had to do that with the `import * from x` syntax since it was throwing deprecated warnings.

sindresorhus commented on package.json:
```diff
@@ -38,27 +42,63 @@
 	},
 	"devDependencies": {
-		"@types/node": "^12.7.4",
+		"@sindresorhus/tsconfig": "^0.4.0",
```
Upgrade this to latest version.

sindresorhus commented on package.json:
```diff
@@ -38,27 +42,63 @@
+		"@types/node": "^12.12.11",
+		"@types/semver": "^6.2.0",
+		"@types/write-file-atomic": "^2.1.2",
+		"@typescript-eslint/eslint-plugin": "^1.13.0",
```
Upgrade this to latest version.

sindresorhus commented on source/index.ts:
```diff
@@ -0,0 +1,746 @@
+import semver = require('semver');
+import onetime from 'onetime';
+
+const plainObject: () => object = () => Object.create(null);
```
Typo

rafaelramalho19 replied:
I don't see any typo in this line Sindre. What did you mean?

sindresorhus replied:
No idea. Must have been a mistake on my part.

sindresorhus commented on source/index.ts:
```diff
@@ -0,0 +1,746 @@
+const INTERNAL_KEY = '__internal__';
+const MIGRATION_KEY = `${INTERNAL_KEY}.migrations.version`;
+
+export default class Conf<T = any> implements Iterable<[string, T]> {
```
```suggestion
export default class Conf<T = unknown> implements Iterable<[string, T]> {
```

sindresorhus commented on source/index.ts:
```diff
@@ -0,0 +1,746 @@
+import Ajv = require('ajv');
+import debounceFn = require('debounce-fn');
+import semver = require('semver');
+import onetime from 'onetime';
```
```suggestion
import onetime = require('onetime');
```

sindresorhus commented on source/index.ts:
```diff
@@ -0,0 +1,746 @@
+
+	_defaultValues: DefaultValues = {};
+
+	_validator?: Ajv.ValidateFunction;
```
This should not be optional.

rafaelramalho19 replied:
In the codebase the validator only exists if there's a schema. Doesn't that make it optional?

Superjo149 replied:
@sindresorhus Looks correct to me as well. It will only exist if `options.schema` is present. The validate method already check its existence.

sindresorhus commented on source/index.ts:
```diff
@@ -0,0 +1,746 @@
+}
+
+export type Serializer = (...args: unknown[]) => string;
+export type Deserializer = (arg: string | Buffer) => unknown;
```
```suggestion
export type Deserializer = (data: string) => unknown;
```

It should not accept `Buffer`.

sindresorhus commented on source/index.ts:
```diff
@@ -0,0 +1,746 @@
+	}
+}
+
+export type Serializer = (...args: unknown[]) => string;
```
It should not accept spreaded arguments. Please make sure to compare against the existing type definition.

sindresorhus commented on source/index.ts:
```diff
@@ -0,0 +1,746 @@
+export type Serializer = (...args: unknown[]) => string;
+export type Deserializer = (arg: string | Buffer) => unknown;
+export type DefaultValues = {
+	[key: string]: object;
```
Use something more specific than `object`.

sindresorhus commented on source/index.ts:
```diff
@@ -0,0 +1,746 @@
+	[key: string]: (store: Conf) => void;
+};
+
+export type Schema = object | boolean;
```
This type is not correct. Look at the existing type definition.

sindresorhus commented on source/index.ts:
```diff
@@ -0,0 +1,746 @@
+		Config used if there are no existing config.
+		** Note: The values in `defaults` will overwrite the `default` key in the `schema` option.
+	*/
+	defaults?: Readonly<object>;
```
Use something more specific than `object`.

sindresorhus commented on source/index.ts:
```diff
@@ -0,0 +1,746 @@
+			// https://github.com/webpack/webpack/issues/196
+			const packageData = packagePath && JSON.parse(fs.readFileSync(packagePath, 'utf8'));
+
+			return packageData || {};
```
```suggestion
			return packageData ?? {};
```

sindresorhus commented on source/index.ts:
```diff
@@ -0,0 +1,746 @@
+		this._validate(store);
+		try {
+			assert.deepEqual(fileStore, store);
+		} catch (_) {
```
```suggestion
		} catch {
```

sindresorhus commented on source/index.ts:
```diff
@@ -0,0 +1,746 @@
+		return true;
+	}
+
+	_get(key: string, defaultValue?: unknown): any {
```
Don't use `any`.

sindresorhus commented on source/index.ts:
```diff
@@ -0,0 +1,746 @@
+	@param key - The key of the item to get.
+	@param defaultValue - The default value if the item does not exist.
+	*/
+	get(key: string, defaultValue?: unknown): unknown {
```
The types here are not correct. It should be strongly typed. Please see existing type definition.

sindresorhus commented on source/index.ts:
```diff
@@ -0,0 +1,746 @@
+	@param key - You can use [dot-notation](https://github.com/sindresorhus/dot-prop) in a key to access nested properties.
+	@param value - Must be JSON serializable. Trying to set the type `undefined`, `function`, or `symbol` will result in a `TypeError`.
+	*/
+	set(key: string | {[key: string]: unknown}, value?: unknown): void {
```
Same as above.

sindresorhus commented on source/index.ts:
```diff
@@ -0,0 +1,746 @@
+
+	@param keys - The keys of the items to reset.
+	*/
+	reset(...keys: string[]): void {
```
Same as above.

rafaelramalho19 replied:
What typing do you suggest here? Because `string` is already equal to `K` (which is just the key of Conf)

sindresorhus replied:
The answer is in the existing TS definition:

https://github.com/sindresorhus/conf/blob/97bcc918a109ba964f7ef8a091cc5a762ae03576/index.d.ts#L289

Yes, both resolve to a string, but having it strongly typed will let editors auto-complete the key names and let TS enforce only valid key names.

rafaelramalho19 replied:
I've not been able to make it work, it always says this:

![image](https://user-images.githubusercontent.com/24696635/74832561-7c1b4c00-530f-11ea-99e6-2d6e64e691cf.png)

Do you have any suggestions for my problem? :/

sindresorhus commented on source/index.ts:
```diff
@@ -0,0 +1,746 @@
+		return () => this.events?.removeListener('change', onChange);
+	}
+
+	encryptData(data: string | Buffer): string | Buffer {
```
```suggestion
	_encryptData(data: string | Buffer): string | Buffer {
```

sindresorhus commented on test/index.ts:
```diff
@@ -808,22 +814,24 @@ test('`watch` option watches for config file changes by file write', async t =>
 
 	const migrations = {
-		'0.0.3': store => {
+		'0.0.3': (store: any) => {
```
Don't use `any`.

sindresorhus commented:
> Try to make sure that the output dist/index.d.ts matches the old index.d.ts as close as possible.

I don't think you have done this thoroughly enough. The types in the existing type definition are better in many places.

rafaelramalho19 commented:
> > Try to make sure that the output dist/index.d.ts matches the old index.d.ts as close as possible.
> 
> I don't think you have done this thoroughly enough. The types in the existing type definition are better in many places.

You're right Sindre. Sorry about that.

I've been a bit busy lately but don't worry about this becoming stale, I'll just pick it up as soon as I can.

sindresorhus commented on package.json:
```diff
@@ -13,12 +14,14 @@
 	"scripts": {
-		"test": "xo && ava && tsd"
+		"build": "del dist && tsc",
+		"test": "npm run build && xo source/** && nyc ava",
```
Use the XO `files` config instead of specifying it here.

rafaelramalho19 replied:
From what I read and tested, the `files` is only used in the `overrides` option. I don't want a override here, I just want the linting to run in the source folder.

Since there's no `includes` or `cwd` in `xo`, I see 2 options:
1. Ignore all folders except `source` with the `ignores` option.
2. Leave this part as it is.

Do you have any other suggestion about this?

sindresorhus replied:
Ah, right. Ok, let's go with 2 then.

sindresorhus commented on source/index.ts:
```diff
@@ -0,0 +1,744 @@
+	@param key - The key of the item to get.
+	@param defaultValue - The default value if the item does not exist.
+	*/
+	get(key: string, defaultValue?: T): T | undefined {
```
This should have an overload that makes the return value `T` when `defaultValue` is given.

So the definitions should be:

```ts
get(key: string): T | undefined {
get(key: string, defaultValue: T): T {
```

rafaelramalho19 replied:
Are you sure about the return type of the second overload?
There's a wide range of inputs to the function where it could return undefined.

one example, assuming the store has: 
`[cool-key]: undefined`

using `get('cool-key', 'this should be the fallback`)` will return `undefined`.

Is also raises the question:
- Do we want undefined, but declared, values in the store to be returned or fallbacked?

sindresorhus replied:
The store cannot have `undefined`:

>The value must be JSON serializable. Trying to set the type undefined, function, or symbol will result in a TypeError.

My suggestion assumes you also do https://github.com/sindresorhus/conf/pull/96#discussion_r362317785 to all the set/get/reset/etc methods, so they're strongly typed.

sindresorhus commented on source/index.ts:
```diff
@@ -0,0 +1,744 @@
+}
+
+export type Migrations<T> = {
+	[key: string]: (store: Conf<T>) => void;
```
```suggestion
	[semverSpecifier: string]: (store: Conf<T>) => void;
```

sindresorhus commented on source/index.ts:
```diff
@@ -0,0 +1,744 @@
+	@param key - The key wo watch.
+	@param callback - A callback function that is called on any changes. When a `key` is first set `oldValue` will be `undefined`, and when a key is deleted `newValue` will be `undefined`.
+	*/
+	onDidChange(key: string, callback: (...args: unknown[]) => void): () => unknown {
```
This should be strongly-typed.

https://github.com/sindresorhus/conf/blob/97bcc918a109ba964f7ef8a091cc5a762ae03576/index.d.ts#L316-L319

sindresorhus commented on source/index.ts:
```diff
@@ -0,0 +1,744 @@
+	/**
+	Delete all items.
+	*/
+	clear(): void{
```
```suggestion
	clear(): void {
```

sindresorhus commented on source/index.ts:
```diff
@@ -0,0 +1,744 @@
+				const initializationVector = data.slice(0, 16);
+				const password = crypto.pbkdf2Sync(this.encryptionKey, initializationVector.toString(), 10000, 32, 'sha512');
+				const decipher = crypto.createDecipheriv(encryptionAlgorithm, password, initializationVector);
+				const slicedData: any = data.slice(17);
```
Don't use `any`.

sindresorhus commented on source/index.ts:
```diff
@@ -0,0 +1,744 @@
+	get store(): {[key: string]: T} {
+		try {
+			const data: string | Buffer = fs.readFileSync(this.path, this.encryptionKey ? null : 'utf8');
+			const dataString = this._encryptData(data) as string;
```
You should not use `as string` here as encryptData could return a Buffer. Generally, stay away from `as` unless necessary as it could easily hide type errors.

sindresorhus commented:
There were still some unresolved inline comments like https://github.com/sindresorhus/conf/pull/96#discussion_r350161217

sindresorhus commented on source/index.ts:
```diff
@@ -0,0 +1,744 @@
+	_migrate(migrations: Migrations<T>, versionToMigrate: string): void {
+		let previousMigratedVersion: string = this._get(MIGRATION_KEY, '0.0.0') as string;
+
+		const newerVersions: string[] = Object.keys(migrations)
```
```suggestion
		const newerVersions = Object.keys(migrations)
```

sindresorhus commented:
@rafaelramalho19 Still interested in finishing this?

rafaelramalho19 commented:
> @rafaelramalho19 Still interested in finishing this?

Yes, I was just a bit busier these few last weeks 😅

rafaelramalho19 commented:
> @rafaelramalho19 Still interested in finishing this?

Maybe someone else can pick up on this? I'm more than happy to let someone else finish this, I guess I'm not experienced enough in Typescript for converting all these code.

Sorry for the long wait Sindre.

sindresorhus commented:
Sure, no worries at all. I totally understand. TypeScript is hard. I think so too.

sindresorhus commented:
Moved to https://github.com/sindresorhus/conf/issues/86.
